cachi2/interface/cli.py

```diff
@@ -132,6 +133,17 @@ def fetch_deps(
         resolve_path=True,
         help="Write output files to this directory.",
     ),
+    backup_source: bool = typer.Option(
+        True,
+        "--backup-source",
```

Comment:
Do we really need this configurable? My understanding is that copying the sources is something cachi2 deeply cares about wrt/ "immutable" dependency fetching and validation, but not necessarily what users care about, IOW the fact we copy trees should be transparent to users. I might be missing some crucial bit of information though, so let me know if ^this is an incorrect premise.

Also, keep in mind that not only direct users of cachi2 might not really care about how we handle project processing internally,but if they use kernel 5.17 or lower with a file system supporting reflinks and the point the backup to a different mount point on the very same file system as the project source, they might encounter some `ReflinkImpossibleError` mumbo jumbo error about unsupported cross device links [1] leaving them in confusion what the real problem is, so I think this whole thing should be handled by us transparently.

[1] https://btrfs.readthedocs.io/en/latest/Reflink.html (I'm sorry I couldn't find a dedicated resource on the 5.17 restriction, so BTRFS docs will have to do to make my point)

Reply:
Yeah, the idea is to make it transparent (that's why we have the default set), but also to allow the user to specify a different path if a corner case arises. We never know how the user environment will look like, so I thought the flexibilty could help.

I do agree it is to cover a very minority of cases, so maybe we could make this a [config flag](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/config.py), or even ditch it for now and add it later if a user requests it.

Reply:
> but also to allow the user to specify a different path if a corner case arises. We never know how the user environment will look like

Well, if the idea is truly making it transparent to the user, I don't think we should make it configurable, because you'd be providing configuration for behaviour that is to remain an internal implementation detail not documented anywhere - otherwise it's not transparent and requires user interaction (with a set default). The fact that we get rid of the copied directory before user request processed suggests the user isn't supposed to interact with this in any way (as they shouldn't!).

What gives us the confidence users will know what to do with `ReflinkImpossibleError` errors if we currently don't know all the corner cases. From that POV it makes sense to me to just model this as:
```
if reflink_supported:
    try:
        copy_with_reflinks()
   except ReflinkImpossibleError:
        fallback to regular copy (let the exception from copy propagate all the way up)
else:
    use regular copy
```

> so maybe we could make this a [config flag](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/config.py)

In my book config options should match CLI options (simply for proper order of precedence wrt/ configuration), so once we add a config option, it should be followed by a CLI options. Nonetheless, either of the two will instantly make this behaviour a public API which we have to be very careful about not to fall into a trap of being stuck with it and not being able to change it easily at our will without breaking CLI logic and use cases (this goes in general, not just for this proposal)

Reply:
Definitely +1 on this being an implementation detail, and not something end users should even know about.

Reply:
Removed all the flags from the CLI and changed the code to fallback to `shutil.copy2` in case a `ReflinkImpossibleError` happens.